src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1807,7 +1807,29 @@ static void ProcessNode(RegexNode node, RegexNode subsequent)
                     // If the node can be changed to atomic based on what comes after it, do so.
                     switch (node.Kind)
                     {
-                        case RegexNodeKind.Oneloop or RegexNodeKind.Notoneloop or RegexNodeKind.Setloop when CanBeMadeAtomic(node, subsequent, allowSubsequentIteration: true):
+                        case RegexNodeKind.Oneloop or RegexNodeKind.Notoneloop or RegexNodeKind.Setloop when CanBeMadeAtomic(node, subsequent, iterateNullableSubsequent: true, allowLazy: false):
+                            // The greedy loop doesn't overlap with what comes after it, which means giving anything it matches back will not
+                            // help the overall match to succeed, which means it can simply become atomic to match as much as possible. The call
+                            // to CanBeMadeAtomic passes iterateNullableSubsequent=true because, in a pattern like a*b*c*, when analyzing a*, we
+                            // want to examine the b* and the c* rather than just giving up after seeing that b* is nullable; in order to make
+                            // the a* atomic, we need to know that anything that could possibly come after the loop doesn't overlap.
+                            node.MakeLoopAtomic();
+                            break;
+
+                        case RegexNodeKind.Onelazy or RegexNodeKind.Notonelazy or RegexNodeKind.Setlazy when CanBeMadeAtomic(node, subsequent, iterateNullableSubsequent: false, allowLazy: true):
+                            // The lazy loop doesn't overlap with what comes after it, which means it needs to match as much as its allowed
+                            // to match in order for there to be a possibility that what comes next matches (if it doesn't match as much
+                            // as it's allowed and there was still more it could match, then what comes next is guaranteed to not match,
+                            // since it doesn't match any of the same things the loop matches).  We don't want to just make the lazy loop
+                            // atomic, as an atomic lazy loop matches as little as possible, not as much as possible.  Instead, we want to
+                            // make the lazy loop into an atomic greedy loop.  Note that when we check CanBeMadeAtomic, we need to set
+                            // "iterateNullableSubsequent" to false so that we only inspect non-nullable subsequent nodes.  For example,
```

Comment:
Thanks for adding the comments, they are very helpful. Should we also add a line in there for why one uses allowLazy=true and the other one false?